ocaml/src/nsm_host_client.ml

```diff
@@ -30,62 +30,85 @@ class type basic_client = object
-    >>= fun () ->
-
-    let res_buf = Llio.make_buffer res_s 0 in
-    let open Nsm_model in
-    match Llio.int_from res_buf with
-    | 0 ->
-      Lwt.return (deserializer res_buf)
-    | ierr ->
-      let ec = Err.int2err ierr in
-      let payload = Llio.string_from res_buf in
-      let exn = Err.Nsm_exn (ec, payload) in
-      Lwt_log.debug_f
-        "nsm host operation %s failed: %i %s,%s"
-        tag_name
-        ierr (Err.show ec) payload
-      >>= fun () ->
-      Lwt.fail exn
-  in
-  let do_request session tag serialize_request deserialize_response =
-    let buf = Buffer.create 20 in
-    Llio.int32_to buf tag;
-    let tag_name = tag_to_name tag in
-    Lwt_log.debug_f "nsm_host_client: %s" tag_name >>= fun () ->
-    serialize_request buf;
-    Lwt_unix.with_timeout
-      10.
-      (fun () ->
-       Lwt_extra2.llio_output_and_flush oc (Buffer.contents buf) >>= fun () ->
-       read_response tag_name deserialize_response
-      )
-  in
-  object(self :# basic_client)
+  Lwt_log.debug_f
+    "nsm host client read response of size %i for %s"
+    (String.length res_s) tag_name
+  >>= fun () ->
 
-    method query : type i o.
-      ?consistency : Consistency.t ->
-      (i, o) query -> i -> o Lwt.t =
-      fun ?(consistency = Consistency.Consistent) command req ->
-      do_request
+  let res_buf = Llio.make_buffer res_s 0 in
+  let open Nsm_model in
+  match Llio.int_from res_buf with
+  | 0 ->
+     Lwt.return (deserializer res_buf)
+  | ierr ->
+     let ec = Err.int2err ierr in
+     let payload = Llio.string_from res_buf in
+     let exn = Err.Nsm_exn (ec, payload) in
+     Lwt_log.debug_f
+       "nsm host operation %s failed: %i %s,%s"
+       tag_name
+       ierr (Err.show ec) payload
+     >>= fun () ->
+     Lwt.fail exn
+
+let _do_request connection (session:Nsm_protocol.Session.t)
+                tag serialize_request deserialize_response
+  =
+  let buf = Buffer.create 20 in
+  let ic,oc = connection in
+  Llio.int32_to buf tag;
+  let tag_name = tag_to_name tag in
+  Lwt_log.debug_f "nsm_host_client: %s" tag_name >>= fun () ->
+  serialize_request buf;
+  Lwt_unix.with_timeout
+    10.
+    (fun () ->
+      Lwt_extra2.llio_output_and_flush oc (Buffer.contents buf) >>= fun () ->
+      _read_response ic tag_name deserialize_response
+    )
+
+
+let _query : type i o.
+             ?consistency : Consistency.t ->
+             Llio.lwtic * Llio.lwtoc ->
+             Nsm_protocol.Session.t ->
+             (i, o) query -> i -> o Lwt.t
+  =
+  fun ?(consistency = Consistency.Consistent)
+      connection session command req ->
+      _do_request
+        connection
         session
         (command_to_tag (Wrap_q command))
         (fun buf ->
          Consistency.to_buffer buf consistency;
          write_query_i command buf req)
         (read_query_o command)
 
+let _get_version connection session =
+  _query connection session ~consistency:Consistency.Consistent GetVersion ()
```

Comment:
Same remark as in albamgr_client: I think this can (and should) be an inconsistent read